tick/eval/eval_binary_node.go

```diff
@@ -0,0 +1,334 @@
+
+	if isDynamicNode(node.Left) || isDynamicNode(node.Right) {
+		expression.evaluationFn = expression.evaluateDynamicNode
+	} else {
+		expression.leftSideType = getConstantNodeType(node.Left)
+		expression.rightSideType = getConstantNodeType(node.Right)
+
+		expression.evaluationFn = expression.findEvaluationFn(expression.leftSideType, expression.rightSideType)
+	}
+
+	return expression, nil
+}
+
+func (n *EvalBinaryNode) GetType(scope ReadOnlyScope, executionState ExecutionState) (ValueType, error) {
+	return findNodeTypes(n.Type, []NodeEvaluator{n.leftSideEvaluator, n.rightSideEvaluator}, scope, executionState)
+}
+
+func (n *EvalBinaryNode) EvalRegex(scope *tick.Scope, executionState ExecutionState) (*regexp.Regexp, error) {
+	return nil, CreateTypeGuardFailedError(TRegex, n.Type)
+}
+
+func (n *EvalBinaryNode) EvalString(scope *tick.Scope, executionState ExecutionState) (string, error) {
+	return "", CreateTypeGuardFailedError(TString, n.Type)
+}
+
+// EvalBool executes the expression based on eval bool
+func (e *EvalBinaryNode) EvalBool(scope *tick.Scope, executionState ExecutionState) (bool, error) {
+	result, err := e.evalWithNodes(scope, executionState, e.leftSideEvaluator, e.rightSideEvaluator)
+	if err != nil {
+		return false, err.error
+	}
+
+	if result.IsBoolValue {
+		return result.BoolValue, nil
+	}
+
+	return false, fmt.Errorf("expression returned unexpected type %T", result.value())
+}
+
+// EvalNum executes the expression based on eval numeric
+func (e *EvalBinaryNode) EvalFloat64(scope *tick.Scope, executionState ExecutionState) (float64, error) {
+	result, err := e.evalWithNodes(scope, executionState, e.leftSideEvaluator, e.rightSideEvaluator)
+	if err != nil {
+		return float64(0), err.error
+	}
+
+	if result.IsFloat64Value {
+		return result.Float64Value, nil
+	}
+
+	if result.IsInt64Value {
+		return float64(0), CreateTypeGuardFailedError(TFloat64, TInt64)
+	}
+
+	return float64(0), CreateTypeGuardFailedError(TFloat64, e.Type)
+}
+
+func (e *EvalBinaryNode) EvalInt64(scope *tick.Scope, executionState ExecutionState) (int64, error) {
+	result, err := e.evalWithNodes(scope, executionState, e.leftSideEvaluator, e.rightSideEvaluator)
+	if err != nil {
+		return int64(0), err.error
+	}
+
+	if result.IsInt64Value {
+		return result.Int64Value, nil
+	}
+
+	if result.IsFloat64Value {
+		return int64(0), CreateTypeGuardFailedError(TInt64, TFloat64)
+	}
+
+	return int64(0), CreateTypeGuardFailedError(TInt64, e.Type)
+}
+
+func (e *EvalBinaryNode) evalWithNodes(scope *tick.Scope, executionState ExecutionState, left, right NodeEvaluator) (resultContainer, *ErrSide) {
+	// TODO: handle errors, for real dynamic values! for example refNode: first is float then int64
+	if e.evaluationFn == nil {
+		err := e.validateEvaluation(scope, executionState, e.operator, left, right)
+		if err != nil {
+			return boolFalseResultContainer, &ErrSide{error: err}
+		}
+	}
+
+	evaluationResult, err := e.evaluationFn(scope, executionState, left, right)
```

Comment:
Done - https://github.com/influxdata/kapacitor/pull/491/commits/6eefb119d7505de0f17cb77c22ac78e30074e235

The historic reason there evalWithNodes because when we used "Value" to determine the type of the node and we want to preserve the state (for function nodes), so I made "constanizeNode(v interface{})".. to sum it up: a mess.